src/database/tuple_desc.cpp

```diff
@@ -1,36 +1,113 @@
 #include "tuple_desc.h"
 
 namespace emerald {
-TupleDesc::TupleDesc TupleDesc::combine(TupleDesc td1, TupleDesc td2) {
-}
-
 TupleDesc::TupleDesc() {
 }
 
-TupleDesc::TupleDesc(Type type_array[]) {
+TupleDesc::TupleDesc(std::vector<Type> type_array) {
+  std::vector<std::string> empty_string = new std::vector<std::string>(0);
+  TupleDesc(type_array, empty_string);
 }
 
-TupleDesc::TupleDesc(Type type_array[], std::string field_array[]) {
+TupleDesc::TupleDesc(std::vector<Type> type_array, std::vector<std::string> field_array) {
+  types = new std::vector<Type>(type_array.size());
+  // memcpy ?
+  for (size_t i = 0; i < type_array.size(); i++) {
+      types.push_back(type_array[i]);
+  }
+  
+  names = new std::vector<std::string>(field_array.size());
+  for (size_t i = 0; i < field_array.size(); i++) {
+      names.push_back(field_array[i]);
+  }
 }
 
 TupleDesc::TupleDesc(const TupleDesc & other) {
 }
 
 int TupleDesc::get_number_fields() const {
+  return types.size();
 }
 
 std::string TupleDesc::get_field_name(int index) const {
+  if (index < 0 || i >= types.size()) {
+    // throw new NoSuchElementException("Index Out of Range."); 
+  } else {
+    return names[i];
+  }
 }
 
-int TupleDesc::get_index(std::string name) const {
+std::string TupleDesc::get_field_type(int index) const {
+  if (index < 0 || i >= types.size()) {
+    // throw new NoSuchElementException("Index Out of Range."); 
+  } else {
+    return types[i];
+  }
 }
 
-Type TupleDesc::get_type(int index) const {
+int TupleDesc::FieldNameToIndex(std::string name) {
+  if (name != NULL) {
+    for (size_t i = 0; i < name.length(); i++) {
+      if (name == names[i]) {
+        return i;
+      }
+    }
+  }
+  // throw new NoSuchElementException("Unknown Field Name.");
 }
 
 int TupleDesc::get_size() const {
+  int size = 0;
+  for (size_t i = 0; i < types.size(); i++) {
+    size += sizeof(types[i]);
+  }
+  return size;
 }
 
-bool TupleDesc::operator==(TupleDesc other) {
+TupleDesc::TupleDesc* TupleDesc::Combine(TupleDesc* td1, TupleDesc* td2) {
+  std::vector<Type> types = new std::vector<Type>(td1->types.size() + td2->types.size());
+  std::vector<std::string> names = new std::vector<std::string>(td1->names.size() + td2->names.size());
+  int idx = 0;
+
+  for (size_t i = 0; i < td1->types.size(); i++) {
+    types[idx] = td1->types[i];
+    names[idx++] = td1->types[i];
+  }
+
+  for (size_t i = 0; i < td2->types.size(); i++) {
+    types[idx] = td2->types[i];
+    names[idx++] = td2->types[i];
+  }
+
+  return new TupleDesc(types, names);
 }
+
+bool TupleDesc::operator==(TupleDesc& other) {
+  if (!(dynamic_cast<TupleDesc*>(other) != nullptr)) {
```

Comment:
We shouldn't need to cast the pointer as the function already takes in a TupleDesc object.

I don't think this clause is necessary.